libraries/AP_KDECAN/AP_KDECAN.cpp

```diff
@@ -53,603 +41,282 @@ const AP_Param::GroupInfo AP_KDECAN::var_info[] = {
-
-                    break;
-                }
-            } else if (now - telemetry_last_request > TELEMETRY_INTERVAL_US) {
-                // broadcast telemetry request frame
-                frame_id_t id = { { .object_address = TELEMETRY_OBJ_ADDR,
-                                  .destination_id = BROADCAST_NODE_ID,
-                                  .source_id = AUTOPILOT_NODE_ID,
-                                  .priority = 0,
-                                  .unused = 0 } };
-
-                AP_HAL::CANFrame frame { (id.value | AP_HAL::CANFrame::FlagEFF), nullptr, 0 };
-                timeout = now + TELEMETRY_TIMEOUT_US;
-                int16_t res = _can_iface->send(frame, timeout, 0);
-                if (res == 1) {
-                    telemetry_last_request = now;
-                } else if (res == 0) {
-                    debug_can(AP_CANManager::LOG_ERROR, "strange buffer full when sending message requesting telemetry");
-                } else {
-                    debug_can(AP_CANManager::LOG_ERROR, "error sending message requesting telemetry, result %d", res);
-                }
+            if (send_packet(ESC_INFO_OBJ_ADDR, BROADCAST_NODE_ID, 100)) {
+                _init.detected_bitmask_ms = now_ms;
             }
         }
-    }
+
+    } // while true
 }
 
-void AP_KDECAN::update()
+bool AP_KDECAN_Driver::send_packet_uint16(const uint8_t address, const uint8_t dest_id, const uint32_t timeout_ms, const uint16_t data)
 {
-    if (_rc_out_sem.take(1)) {
-        for (uint8_t i = 0; i < KDECAN_MAX_NUM_ESCS; i++) {
-            if ((_esc_present_bitmask & (1 << i)) == 0) {
-                _scaled_output[i] = 0;
-                continue;
-            }
-            _scaled_output[i] = SRV_Channels::srv_channel(i)->get_output_pwm();
-        }
-
-        _rc_out_sem.give();
-        _new_output.store(true, std::memory_order_release);
-    } else {
-        debug_can(AP_CANManager::LOG_DEBUG, "failed to get PWM semaphore on write");
-    }
+    const uint16_t data_be16 = htobe16(data);
+    return send_packet(address, dest_id, timeout_ms, (uint8_t*)&data_be16, 2);
 }
 
-bool AP_KDECAN::pre_arm_check(char* reason, uint8_t reason_len)
+bool AP_KDECAN_Driver::send_packet(const uint8_t address, const uint8_t dest_id, const uint32_t timeout_ms, const uint8_t *data, const uint8_t data_len)
 {
-    if (!_enum_sem.take(1)) {
-        snprintf(reason, reason_len ,"enumeration state unknown");
-        return false;
-    }
-
-    if (_enumeration_state != ENUMERATION_STOPPED) {
-        snprintf(reason, reason_len, "enumeration running");
-        _enum_sem.give();
-        return false;
-    }
-
-    _enum_sem.give();
+    // broadcast telemetry request frame
+    frame_id_t id = { { .object_address = address,
+                        .destination_id = dest_id,
+                        .source_id = AUTOPILOT_NODE_ID,
+                        .priority = 0,
+                        .unused = 0 } };
```

Comment:
```suggestion
    const frame_id_t id {
        {
            .object_address = address,
            .destination_id = dest_id,
            .source_id = AUTOPILOT_NODE_ID,
            .priority = 0,
            .unused = 0
        }
    };
```

(stop creating on stack and assigning)